odl/solvers/advanced/chambolle_pock.py

```diff
@@ -0,0 +1,278 @@
+
+# External
+
+# Internal
+from odl.operator.operator import Operator
+from odl.operator.pspace_ops import ProductSpaceOperator
+
+__all__ = ('chambolle_pock_solver', 'f_cc_prox_l2_tv', 'g_prox_none')
+
+
+# TODO: add dual gap as convergence measure
+# TODO: diagonal preconditioning
+# TODO: add positivity constraint
+
+def chambolle_pock_solver(K, f_cc_prox, g_prox, sigma, tau, theta=1,
+                          niter=100, partial=None):
+    """Chambolle-Pock aglgorithms for convex optimization problems.
+
+    First order primal-dual hybrid-gradient (PDHG) method for non-smooth
+    convex optimization probelms with known saddle-point structure developed
+    by `Chambolle and Pock`_ (CP).
+
+
+    Parameters
+    ----------
+    K : `Operator` or `ProductSpaceOperator`
+    f_cc_prox : `function`
+       Factory function which, evaluated at ``sigma``, returns the proximal
+       operator of F^*. Domain and range of the returned proximal operator
+       is the range of ``K``.
+    g_prox : `function`
+        Factory function which, evaluated at ``tau``, returns the proximal
+       operator of ``G``. Domain and range of the returned proximal operator
+       is the domain of ``K``.
+    sigma : positive `float`
+        Convergence proofed for ``||K||_2^2 * sigma * tau < 1``
+    tau : positive `float`
+        Convergence proofed for ``||K||_2^2 * sigma * tau < 1``
+    theta : `float` in [0, 1], optional
+        Relaxation paramter
+    niter : `int`, optional
+        Number of iterations
+    partial : `Partial`, optional
+        If not `None` the object passed executes code per iteration,
+        e.g. plotting each iterate
+
+    Returns
+    -------
+    x : `DiscreteLpVector`
+
+    References
+    ----------
+    Original paper by `Chambolle and Pock`_. Also see `diagonal
+    preconditioning`_ techniques. This implementation is based on the
+    article on convex optimization problem prototyping for image
+    reconstruction in computed tomography by `Sidky et al.`_.
+
+    .. _Chambolle and Pock: http://dx.doi.org/10.1007/s10851-010-0251-1
+    .. _diagonal preconditioning: http://dx.doi.org/10.1109/ICCV.2011.6126441
+    .. _Sidky et al.: http://stacks.iop.org/0031-9155/57/i=10/a=3065
+    """
+    # Initialize the primal and dual variable
+    x = K.domain.zero()
+    y = K.range.zero()
+    # Relaxation variable
+    xbar = x.copy()
+
+    # Initialize the proximal operator of the convex conjugate of functional F
+    f_cc_prox_sigma = f_cc_prox(sigma)
+    # Initialize the proximal operator of functional G
+    g_prox_tau = g_prox(tau)
+    # Adjoint of the product space operator
+    Kadjoint = K.adjoint
+
+    for nn in range(niter):
+        # Copy required for relaxation
+        xold = x.copy()
+        # Gradient descent in the dual variable y
+        f_cc_prox_sigma(y + sigma * K(xbar), out=y)
+        # Gradient descent in the primal variable x
+        g_prox_tau(x - tau * Kadjoint(y), out=x)
```

Comment:
Could be optimized somewhat by 

`g_prox_tau(x + (- tau) * Kadjoint(y), out=x)`